### PR TITLE
Support remaining longhand flexbox style properties for servo

### PR DESCRIPTION
**Pull request by stshine**
- [x] `./mach build -d` does not report any errors
- [X] `./mach test-tidy --faster` does not report any errors
- [X] There are tests for these changes

Add the remaining longhand CSS properties of flexbox for servo,
including `flex-wrap`, `flex-grow`, `flex-shrink`, `justify-content`,
`align-items`, `align-self` and `align-content`.

Tests for the remaining longhand properties are also enabled. Test for
default value of `align-self` is modified because the new default value
for this property is "auto" is current flexbox CR.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11417)

### CODE REVIEW

File: tests/wpt/metadata-css/css-flexbox-1_dev/html/flexbox_computedstyle_flex-flow-nowrap.htm.ini [hunk 1]
```diff
@@ -1,3 +1,3 @@
 [flexbox_computedstyle_flex-flow-nowrap.htm]
   type: reftest
-  expected: FAIL
+  expected: PASS
```
Comment (ConnorGBrewster) [inline tests/wpt/metadata-css/css-flexbox-1_dev/html/flexbox_computedstyle_flex-flow-nowrap.htm.ini]: Any tests that are expected to pass, can be removed from the `ini`.

Comment (stshine) [inline tests/wpt/metadata-css/css-flexbox-1_dev/html/flexbox_computedstyle_flex-flow-nowrap.htm.ini]: Well this test is for the shorthand property `flex-grow`, which does not exist while all its longhand properties, `flex-direction` and `flex-wrap`, are set to their default values. IMO maybe those passed tests should be removed from `ini` after these shorthand properties are actually added.

Comment (KiChjang) [inline tests/wpt/metadata-css/css-flexbox-1_dev/html/flexbox_computedstyle_flex-flow-nowrap.htm.ini]: I think that's what @ConnorGBrewster is trying to say, tests that are passing can has its ini file removed.

**Comment by mbrubeck** [comment]
-S-awaiting-review +S-needs-code-changes 

<details><summary><i>Previously, highfive wrote…</i></summary>
> warning **Warning** warning
> 
> * This pull request modifies the contents of
> 'tests/wpt/css-tests/', which are overwriten occasionally whenever the
> directory is synced from upstream.
</details>


---

Reviewed 42 of 42 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[tests/wpt/css-tests/css-flexbox-1_dev/html/flexbox_computedstyle_align-self-invalid.htm, line 20 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11417#-KIc_O1yDLpZO0kMHToT:-KIc_O2-IarN2-wuK5dB:1544080747) ([raw file](https://github.com/servo/servo/blob/a7733b38d40ff074a0d087597be4fdf3f216c517/tests/wpt/css-tests/css-flexbox-1_dev/html/flexbox_computedstyle_align-self-invalid.htm#L20)):*
I don't think the computed value should be `auto`.  According to https://drafts.csswg.org/css-align/#valdef-align-self-auto

> The `auto` keyword computes to the computed value of `align-items` on the parent or `normal` if the box has no parent.

and the computed value of `align-items` on the parent in this case should be the initial value, `normal`.

---

*[tests/wpt/metadata-css/css-flexbox-1_dev/html/flexbox_computedstyle_flex-flow-nowrap.htm.ini, line 3 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11417#-KIcCwo2IDkJjjB6YR-2-r1-3:-KIc_zrj9oO8AUqcYxv4:407389442) ([raw file](https://github.com/servo/servo/blob/a7733b38d40ff074a0d087597be4fdf3f216c517/tests/wpt/metadata-css/css-flexbox-1_dev/html/flexbox_computedstyle_flex-flow-nowrap.htm.ini#L3)):*
<details><summary><i>Previously, KiChjang (Keith Yeung) wrote…</i></summary>
> I think that's what @ConnorGBrewster is trying to say, tests that are passing can has its ini file removed.
</details>
Yes, please remove any ini files that set "expected: PASS". `PASS` is the default value, so these have no effect.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11417#-:-KIc_p7NLV9TeWOKuLLg:1458266033)*

**Comment by mbrubeck** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[components/style/properties/longhand/position.mako.rs, line 64 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11417#-KIccZs1d6E-XRmAUs2P:-KIccZs2kKFFS69mIzeE:-1196550796) ([raw file](https://github.com/servo/servo/blob/52f17a88141c8a55de3816e1a7a169cced518b36/components/style/properties/longhand/position.mako.rs#L64)):*
> ```Rust
> 
> // Flex container properties
> ${helpers.single_keyword("flex-direction", "row row-reverse column column-reverse", experimental=True)}
> ```

Why is this no longer marked "experimental"?

---

*[components/style/properties/longhand/position.mako.rs, line 66 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11417#-KIcd2fpYpISg-a790gA:-KIcd2fpYpISg-a790gB:-984657690) ([raw file](https://github.com/servo/servo/blob/a7733b38d40ff074a0d087597be4fdf3f216c517/components/style/properties/longhand/position.mako.rs#L66)):*
> ```Rust
> ${helpers.single_keyword("flex-direction", "row row-reverse column column-reverse")}
> 
> ${helpers.single_keyword("flex-wrap", "nowrap wrap wrap-reverse")}
> ```

These should all be marked `products="gecko"` or `experimental=True` until Servo layout has implemented them.

If they are `experimental` then you can enable them in testing by adding a corresponding preference to `resources/prefs.json` and `tests/wpt/metadata-css/css-flexbox-1_dev/__dir__.ini`.  You can also enable the preference on the Servo command line using the `--pref` option.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11417)*

**Comment by stshine** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[components/style/properties/longhand/position.mako.rs, line 66 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11417#-KIcd2fpYpISg-a790gA:-KIcjfZDYa3LL3NLo7Oz:-1751321389) ([raw file](https://github.com/servo/servo/blob/a7733b38d40ff074a0d087597be4fdf3f216c517/components/style/properties/longhand/position.mako.rs#L66)):*
<details><summary><i>Previously, mbrubeck (Matt Brubeck) wrote…</i></summary>
> These should all be marked `products="gecko"` or `experimental=True` until Servo layout has implemented them.
> 
> If they are `experimental` then you can enable them in testing by adding a corresponding preference to `resources/prefs.json` and `tests/wpt/metadata-css/css-flexbox-1_dev/__dir__.ini`.  You can also enable the preference on the Servo command line using the `--pref` option.
</details>
Oh I will mark them as experimental. I am trying on flexbox layout.

---

*[tests/wpt/css-tests/css-flexbox-1_dev/html/flexbox_computedstyle_align-self-invalid.htm, line 20 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11417#-KIc_O1yDLpZO0kMHToT:-KIcmBT0ucBQDfXO_unz:-2112973229) ([raw file](https://github.com/servo/servo/blob/a7733b38d40ff074a0d087597be4fdf3f216c517/tests/wpt/css-tests/css-flexbox-1_dev/html/flexbox_computedstyle_align-self-invalid.htm#L20)):*
<details><summary><i>Previously, mbrubeck (Matt Brubeck) wrote…</i></summary>
> I don't think the computed value should be `auto`.  According to https://drafts.csswg.org/css-align/#valdef-align-self-auto
> 
> > The `auto` keyword computes to the computed value of `align-items` on the parent or `normal` if the box has no parent.
> 
> and the computed value of `align-items` on the parent in this case should be the initial value, `normal`.
</details>
Oh I failed to notice that these properties are not only for flexbox layout. This makes the situation much more complex, and I will read this spec first.

---

*[tests/wpt/metadata-css/css-flexbox-1_dev/html/flexbox_computedstyle_flex-flow-nowrap.htm.ini, line 3 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11417#-KIcCwo2IDkJjjB6YR-2-r1-3:-KIck5FjGSx9BSedk6-W:-1528396111) ([raw file](https://github.com/servo/servo/blob/a7733b38d40ff074a0d087597be4fdf3f216c517/tests/wpt/metadata-css/css-flexbox-1_dev/html/flexbox_computedstyle_flex-flow-nowrap.htm.ini#L3)):*
<details><summary><i>Previously, mbrubeck (Matt Brubeck) wrote…</i></summary>
> Yes, please remove any ini files that set "expected: PASS". `PASS` is the default value, so these have no effect.
</details>
OK I'll do it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11417)*

**Comment by stshine** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[tests/wpt/css-tests/css-flexbox-1_dev/html/flexbox_computedstyle_align-self-invalid.htm, line 20 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11417#-KIc_O1yDLpZO0kMHToT:-KIjav6atKSzoXUnsF0b:828242681) ([raw file](https://github.com/servo/servo/blob/a7733b38d40ff074a0d087597be4fdf3f216c517/tests/wpt/css-tests/css-flexbox-1_dev/html/flexbox_computedstyle_align-self-invalid.htm#L20)):*
<details><summary><i>Previously, stshine wrote…</i></summary>
> Oh I failed to notice that these properties are not only for flexbox layout. This makes the situation much more complex, and I will read this spec first.
</details>
the [css-align] is a rather big spec and a lot of issues needs to be addressed , and no part of servo needs it right now, and it breaks some tests of flexbox. Would  it be better to implement it later separately?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11417)*

**Comment by stshine** [comment]
Review status: 34 of 45 files reviewed at latest revision, 4 unresolved discussions.

---

*[tests/wpt/css-tests/css-flexbox-1_dev/html/flexbox_computedstyle_align-self-invalid.htm, line 20 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11417#-KIc_O1yDLpZO0kMHToT:-KIjuCwU1P4Nl2LrYYQt:109541126) ([raw file](https://github.com/servo/servo/blob/a7733b38d40ff074a0d087597be4fdf3f216c517/tests/wpt/css-tests/css-flexbox-1_dev/html/flexbox_computedstyle_align-self-invalid.htm#L20)):*
<details><summary><i>Previously, stshine wrote…</i></summary>
> the [css-align] is a rather big spec and a lot of issues needs to be addressed , and no part of servo needs it right now, and it breaks some tests of flexbox. Would  it be better to implement it later separately?
</details>
I have pushed a fix, the 'auto' value of `align-self` should be computed to `align-items` of its parent. Currently the issue is that the flexbox spec indicates that the 'auto' value of `align-self` on an absolutely positioned element should compute to itself while in css-align spec it should never computed to 'auto'. Another issue is that the test flex-flexitem-percentage-prescation passes unexpectedly while it looks totally different from its reference when I run them manually.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11417)*

**Comment by jdm** [comment]
@stshine Did you enable the layout.flex preferences in prefs.json when running manually?

**Comment by stshine** [comment]
@jdm Yes you are right, they are identical now

**Comment by mbrubeck** [comment]
-S-awaiting-review +S-awaiting-answer 

<details><summary><i>Previously, stshine wrote…</i></summary>
> @jdm Yes you are right, they are identical now
</details>


---

Reviewed 11 of 11 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[components/style/properties/properties.mako.rs, line 1904 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11417#-KIn8h7IMhkDjNrYJ7ix:-KIn8h7JRU7JjmEvvvyo:727052311) ([raw file](https://github.com/servo/servo/blob/df20d4f7e7854eff9374391c84f32bf870dac600/components/style/properties/properties.mako.rs#L1904)):*
> ```Rust
>         use computed_values::align_self::T as align_self;
>         use computed_values::align_items::T as align_items;
>         if style.get_position().clone_align_self() == computed_values::align_self::T::auto && !positioned {
> ```

Why is this different for positioned boxes?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11417#-:-KIn8oj7sPTS2OqDSp_3:453359453)*

**Comment by stshine** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[components/style/properties/properties.mako.rs, line 1904 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11417#-KIn8h7IMhkDjNrYJ7ix:-KInPvEJrCjVQkV45SWi:1173530627) ([raw file](https://github.com/servo/servo/blob/df20d4f7e7854eff9374391c84f32bf870dac600/components/style/properties/properties.mako.rs#L1904)):*
<details><summary><i>Previously, mbrubeck (Matt Brubeck) wrote…</i></summary>
> Why is this different for positioned boxes?
</details>
As the flexbox spec indicated: https://drafts.csswg.org/css-flexbox-1/#propdef-align-self

On absolutely positioned elements, a value of auto computes to itself. On all other elements, a value of auto for align-self computes to the value of align-items on the element’s parent, or stretch if the element has no parent. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11417)*

**Comment by mbrubeck** [comment]
-S-awaiting-answer 

@bors-servo r+

<details><summary><i>Previously, mbrubeck (Matt Brubeck) wrote…</i></summary>
> -S-awaiting-review +S-awaiting-answer 
</details>


---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[components/style/properties/properties.mako.rs, line 1904 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11417#-KIn8h7IMhkDjNrYJ7ix:-KInTC2HJX3R2NEoQdzF:1327159541) ([raw file](https://github.com/servo/servo/blob/df20d4f7e7854eff9374391c84f32bf870dac600/components/style/properties/properties.mako.rs#L1904)):*
<details><summary><i>Previously, stshine wrote…</i></summary>
> As the flexbox spec indicated: https://drafts.csswg.org/css-flexbox-1/#propdef-align-self
> 
> On absolutely positioned elements, a value of auto computes to itself. On all other elements, a value of auto for align-self computes to the value of align-items on the element’s parent, or stretch if the element has no parent. 
</details>
Oh, right. You mentioned that before, but then I confused myself by reading different and conflicting versions of the specs. :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11417#-:-KInU1cNiWBXTtf-ER5m:-1014847198)*

**Comment by mbrubeck** [comment]
@bors-servo r+

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.
</details>


---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11417#-:-KIoi-0TFqg_cN6ofcYu:1236647940)*

**Comment by mbrubeck** [comment]
@bors-servo r- delegate+

Found a build error in local testing.

**Comment by stshine** [comment]
It look like that those  NS_STYLE_JUSTIFY_* are not generated in   ports/geckolib/gecko_bindings/structs.rs. Something related to @bholley ?

**Comment by stshine** [comment]
It really took me sometime to figure things out. What confused me most is that the compiling of geckolib seems not to be consistent :-( 
There is two parts of the problem. First is that the NS_STYLE_JUSTIFY_* in [nsStyleConsts.h](https://dxr.mozilla.org/mozilla-central/source/layout/style/nsStyleConsts.h#512) is not generated in structs.rs. Since this file is not totally automatically generated, I may added them manually. On the other hand, when gecko [implemented](https://bugzilla.mozilla.org/show_bug.cgi?id=1176782) [css-align] and [css-grid], The type of `mAlignContent` and `mJustifyContent` was [changed](https://bugzilla.mozilla.org/page.cgi?id=splinter.html&bug=1176782&attachment=8678581) to `uint16_t` and the higher bytes of the type is used as fallback value as the [css-align stated](https://drafts.csswg.org/css-align/#fallback-alignment), while in ports/geckolib/properties.mako.rs the helper [impl_keyword_setter](https://github.com/servo/servo/blob/master/ports/geckolib/properties.mako.rs#L170) assume that all the properties has type `uint8_t`.
I may write special template for `align-content` and `justify-content` to make their setter return `u16`, but that wound not fit into the logic gecko handles them. So I decide to disable these two properties in geckolib for now, and add a fixme for them.

**Comment by stshine** [comment]
Add these experimental attributes to css-flexbox-1_dev/__dir__.ini.

**Comment by stshine** [comment]
Well these two tests didn't pass when I first opened this pull request until I pushed the fixup.
But now I just checked out the first commit and compiled it, these tests get passed too, while they failed on the build server. Can't understand how it could happend...

**Comment by stshine** [comment]
I dumped screenshots in the log,  The option -Z disable-text-aa seems did not take effect, so that the tests on the build server generate text that overflowed its block, while on my computer it didn't because of the antialias settings. On  the other hand, the  display order of the elements of the tests and their references are different, so they generated different pixels.  FAIL should be the expected behavior.

**Comment by stshine** [comment]
It feels really good to have a pull request accepted for the first time. Thank you @mbrubeck and you kind servo people!

**Comment by mbrubeck** [comment]
Good work!

**Pull request merged**